Add collaborators popover menu

shekharnwagh opened the pull request:
# Description

Adds a new collaborators popover that displays all active users in a clickable list which is opened on clicking the collaborators menu with avatars in the top editor bar. Users can click on any collaborator to scroll to their cursor position in the editor. The popover replaces the previous sidebar collaborators list section which has now been removed.

Added a cursor registry to store clientId and cursor ref mappings to facilitate scrolling to cursor positions. This uses a singleton pattern instead of React Context as context propagation was breaking between `EditorPresence` and `BlockCanvasCover.Fill`.

User emails require 'edit' context to be accessible via the REST API which required additional `getUser` call in addition to `getCurrentUser` which only returns the data with `view` context

## Screenshots

| <img width="1424" height="858" alt="Screenshot 2025-11-03 at 9 19 27 PM" src="https://github.com/user-attachments/assets/b6223042-b2a1-468b-b9d1-feab3fb6b334" /> |
| :---: |
| _List of active collaborators, including names and email addresses opened on clicking menu button in top bar_ |

### Video

https://github.com/user-attachments/assets/9122fd24-eae8-4883-b4aa-40b5f51949f9

shekharnwagh committed:
Add hover popover showing all active collaborators

Hovering over the avatar group now shows a popover with all active
users and their details using a new CollaboratorsList component to
display the full list.

chriszarate commented on src/components/avatars.tsx:
```diff
@@ -30,7 +37,11 @@ export function Avatars() {
+		<div
+			className="vip-real-time-collaboration-avatars-container"
+			onMouseEnter={ () => setIsPopoverVisible( true ) }
+			onMouseLeave={ () => setIsPopoverVisible( false ) }
```
In order to be accessible, the popover must be triggerable from a keyboard-focusable element, typically a link, `<button>`, or element with `role="button"`.

shekharnwagh committed:
update styles to align with designs

shekharnwagh committed:
fetch user emails using resolveSelect

shekharnwagh committed:
use button for avatars in top bar

shekharnwagh committed:
Show collaborators popover on click instead of hover

chriszarate commented on src/utilities/entity.ts:
```diff
@@ -1,18 +1,29 @@
+
+async function getUserEmail( userId: number ): Promise< string | undefined > {
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
+	const user = ( await resolveSelect( coreStore ).getUser( userId, {
+		context: 'edit',
+	} ) ) as User< 'edit' > | undefined;
```
Non-admins will not have permission for this request (for other users)

shekharnwagh replied:
We only call this for current user's ID, so wouldn't that be safe ?

chriszarate replied:
Ah yeah, I thought this was being called from collaborator list. In that case we should just use this select instead of getCurrentUser.

shekharnwagh replied:
This doesn't work with user id as `me`, so we need to fetch the current user id first with `getCurrentUser` and then use the id in next request.

chriszarate replied:
Fun, carry on. I can't even find where the `getUser` selector is defined. 🫤

shekharnwagh replied:
They dynamically generated here for entitites using [entitySelectors](https://github.com/WordPress/gutenberg/blob/c54fa0beb059a2e3b2d2f5a32f26ab47598be0b6/packages/core-data/src/index.js#L35-L58) method which is used when creating the store [here](https://github.com/WordPress/gutenberg/blob/c54fa0beb059a2e3b2d2f5a32f26ab47598be0b6/packages/core-data/src/index.js#L121).

Its also mentioned in the `core-data` package readme - https://github.com/WordPress/gutenberg/tree/c54fa0beb059a2e3b2d2f5a32f26ab47598be0b6/packages/core-data#plural

chriszarate replied:
Yay, thanks for the pointers! Dynamically named functions, what a foil for my `ripgrep`ing. :)

chriszarate commented on src/utilities/entity.ts:
```diff
@@ -1,18 +1,29 @@
 export async function getCurrentUserInfo(): Promise< WordPressUserInfo > {
-	const { avatar_urls: avatarUrls, id, name, slug } = select( coreStore ).getCurrentUser() ?? {};
+	const currentUser = select( coreStore ).getCurrentUser() ?? {};
+	const { avatar_urls: avatarUrls, id, name, slug } = currentUser;
```
You should be able to destructure `email` from here. That way each user supplies their own info.

shekharnwagh replied:
I tried that it was emtpy. The return type of getCurrentUser is `User< view >` which doesn't have `email` - https://github.com/WordPress/gutenberg/blob/c54fa0beb059a2e3b2d2f5a32f26ab47598be0b6/packages/core-data/src/entity-types/user.ts#L36-L39

shekharnwagh committed:
Add close button to collaborators popover

shekharnwagh committed:
remove collaborators section from settings panel

The collaborators list is now displayed in the top bar via the
Avatars component with a popover, making the settings panel section
redundant

shekharnwagh committed:
add click to scroll to collaborator cursor

Clicking a collaborator in the list now scrolls to their cursor
position in the editor with a smooth animation and highlight effect.
This helps users quickly find where other collaborators are working.

Implementation uses a singleton cursor registry pattern instead of
React Context due to WordPress Slot/Fill breaking context propagation
between EditorPresence and BlockCanvasCover.Fill components.

shekharnwagh committed:
always hide current user from collaborators list

The collaborators list and cursor rendering now always exclude the
current user. This aligns with how it works in Google Docs. It
simplifies the UI where we don't have to check if current user has
`show my awareness` enabled or not.

shekharnwagh commented on src/components/collaborators-list.tsx:
```diff
@@ -0,0 +1,102 @@
+		// Optionally close the popover after successful scroll
+		if ( success ) {
+			setIsPopoverVisible( false );
+		}
```
I am not entirely sure if we should close the popover when we click on a collaborator. Only added it to de-clutter it when we scroll, by happy to keep the popover open even after click.

maxschmeling replied:
I could make an argument either way on this, so I'm fine letting it close for now.

maxschmeling commented on src/contexts/cursor-registry-context.tsx:
```diff
@@ -0,0 +1,93 @@
+		// Remove each cursor element using stored refs
+		this.cursorMap.forEach( element => {
+			element.remove();
+		} );
```
```suggestion
		this.cursorMap.forEach( element => element.remove() );
```

shekharnwagh committed:
lint changes

Co-authored-by: Max Schmeling <max.schmeling@automattic.com>

shekharnwagh committed:
add a11y labels for collaborators list toggle

shekharnwagh committed:
use native button elements for collaborator list items

shekharnwagh committed:
align outer border radius with that of avatars

shekharnwagh committed:
remove aria-labels for avatars

shekharnwagh committed:
announce scroll to cursor on click

maxschmeling reviewed (approved):
This is working pretty well.

We should improve the accessibility by allowing keyboard navigation through the list (up/down arrows), and as @shekharnwagh pointed out to me we'll probably want some feedback for users in the list that don't have a cursor position to scroll to. But this is a great first pass to get some feedback.
